File: packages/python/elizaos/features/basic_capabilities/services/embedding.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import ModelType, Service, ServiceType

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime


class EmbeddingService(Service):
    name = "embedding"
    service_type = ServiceType.UNKNOWN

    @property
    def capability_description(self) -> str:
        return "Text embedding service for generating and caching text embeddings."

    def __init__(self) -> None:
        self._runtime: IAgentRuntime | None = None
        self._cache: dict[str, list[float]] = {}
        self._cache_enabled: bool = True
        self._max_cache_size: int = 1000

    @classmethod
    async def start(cls, runtime: IAgentRuntime) -> EmbeddingService:
        service = cls()
        service._runtime = runtime
        runtime.logger.info(
            "Embedding service started",
            src="service:embedding",
            agentId=str(runtime.agent_id),
        )
        return service

    async def stop(self) -> None:
        if self._runtime:
            self._runtime.logger.info(
                "Embedding service stopped",
                src="service:embedding",
                agentId=str(self._runtime.agent_id),
            )
        self._cache.clear()
        self._runtime = None

    async def embed(self, text: str) -> list[float]:
        if self._runtime is None:
            raise ValueError("Embedding service not started - no runtime available")

        if self._cache_enabled and text in self._cache:
            # Move to end for LRU behavior
            self._cache[text] = self._cache.pop(text)
            return self._cache[text]

        embedding = await self._runtime.use_model(
            ModelType.TEXT_EMBEDDING,
            text=text,
        )

        if not isinstance(embedding, list):
            raise ValueError(f"Expected list for embedding, got {type(embedding)}")

        embedding = [float(x) for x in embedding]

        if self._cache_enabled:
            self._add_to_cache(text, embedding)

        return embedding

    async def embed_batch(self, texts: list[str]) -> list[list[float]]:
        embeddings: list[list[float]] = []
        for text in texts:
            embedding = await self.embed(text)
            embeddings.append(embedding)
        return embeddings

    def _add_to_cache(self, text: str, embedding: list[float]) -> None:
        if len(self._cache) >= self._max_cache_size:
            oldest_key = next(iter(self._cache))
            del self._cache[oldest_key]
        self._cache[text] = embedding

    def clear_cache(self) -> None:
        self._cache.clear()

    def set_cache_enabled(self, enabled: bool) -> None:
        self._cache_enabled = enabled
        if not enabled:
            self._cache.clear()

    def set_max_cache_size(self, size: int) -> None:
        if size <= 0:
            raise ValueError("Cache size must be positive")
        self._max_cache_size = size
        while len(self._cache) > self._max_cache_size:
            oldest_key = next(iter(self._cache))
            del self._cache[oldest_key]

    async def similarity(self, text1: str, text2: str) -> float:
        embedding1 = await self.embed(text1)
        embedding2 = await self.embed(text2)

        dot_product = sum(a * b for a, b in zip(embedding1, embedding2, strict=True))
        magnitude1 = sum(a * a for a in embedding1) ** 0.5
        magnitude2 = sum(b * b for b in embedding2) ** 0.5

        if magnitude1 == 0 or magnitude2 == 0:
            return 0.0

        return dot_product / (magnitude1 * magnitude2)

File: packages/rust/src/features/basic_capabilities/bootstrap/services/embedding.rs
//! Embedding service implementation.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;

use crate::error::{PluginError, PluginResult};
use crate::runtime::{IAgentRuntime, ModelParams};
use crate::types::ModelType;

use super::{Service, ServiceType};

/// Service for generating text embeddings.
pub struct EmbeddingService {
    runtime: Option<Arc<dyn IAgentRuntime>>,
    cache: HashMap<String, Vec<f32>>,
    cache_enabled: bool,
    max_cache_size: usize,
}

impl EmbeddingService {
    /// Create a new embedding service.
    pub fn new() -> Self {
        Self {
            runtime: None,
            cache: HashMap::new(),
            cache_enabled: true,
            max_cache_size: 1000,
        }
    }

    /// Generate an embedding for the given text.
    pub async fn embed(&mut self, text: &str) -> PluginResult<Vec<f32>> {
        let runtime = self
            .runtime
            .as_ref()
            .ok_or_else(|| PluginError::ServiceNotStarted("embedding".to_string()))?;

        // Check cache first
        if self.cache_enabled {
            if let Some(cached) = self.cache.get(text) {
                return Ok(cached.clone());
            }
        }

        // Generate embedding
        let output = runtime
            .use_model(ModelType::TextEmbedding, ModelParams::with_text(text))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let embedding = output
            .as_embedding()
            .ok_or_else(|| PluginError::ModelError("Expected embedding output".to_string()))?
            .to_vec();

        // Cache result
        if self.cache_enabled {
            self.add_to_cache(text.to_string(), embedding.clone());
        }

        Ok(embedding)
    }

    /// Generate embeddings for multiple texts.
    pub async fn embed_batch(&mut self, texts: &[String]) -> PluginResult<Vec<Vec<f32>>> {
        let mut embeddings = Vec::with_capacity(texts.len());
        for text in texts {
            embeddings.push(self.embed(text).await?);
        }
        Ok(embeddings)
    }

    /// Calculate cosine similarity between two texts.
    pub async fn similarity(&mut self, text1: &str, text2: &str) -> PluginResult<f32> {
        let embedding1 = self.embed(text1).await?;
        let embedding2 = self.embed(text2).await?;

        let dot_product: f32 = embedding1
            .iter()
            .zip(embedding2.iter())
            .map(|(a, b)| a * b)
            .sum();

        let magnitude1: f32 = embedding1.iter().map(|a| a * a).sum::<f32>().sqrt();
        let magnitude2: f32 = embedding2.iter().map(|b| b * b).sum::<f32>().sqrt();

        if magnitude1 == 0.0 || magnitude2 == 0.0 {
            return Ok(0.0);
        }

        Ok(dot_product / (magnitude1 * magnitude2))
    }

    /// Clear the embedding cache.
    pub fn clear_cache(&mut self) {
        self.cache.clear();
    }

    /// Enable or disable caching.
    pub fn set_cache_enabled(&mut self, enabled: bool) {
        self.cache_enabled = enabled;
        if !enabled {
            self.cache.clear();
        }
    }

    /// Set the maximum cache size.
    pub fn set_max_cache_size(&mut self, size: usize) {
        self.max_cache_size = size;
        self.trim_cache();
    }

    fn add_to_cache(&mut self, text: String, embedding: Vec<f32>) {
        if self.cache.len() >= self.max_cache_size {
            // Remove oldest entry (first key)
            if let Some(key) = self.cache.keys().next().cloned() {
                self.cache.remove(&key);
            }
        }
        self.cache.insert(text, embedding);
    }

    fn trim_cache(&mut self) {
        while self.cache.len() > self.max_cache_size {
            if let Some(key) = self.cache.keys().next().cloned() {
                self.cache.remove(&key);
            }
        }
    }
}

impl Default for EmbeddingService {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl Service for EmbeddingService {
    fn name(&self) -> &'static str {
        "embedding"
    }

    fn service_type(&self) -> ServiceType {
        ServiceType::Core
    }

    async fn start(&mut self, runtime: Arc<dyn IAgentRuntime>) -> PluginResult<()> {
        runtime.log_info("service:embedding", "Embedding service started");
        self.runtime = Some(runtime);
        Ok(())
    }

    async fn stop(&mut self) -> PluginResult<()> {
        if let Some(runtime) = &self.runtime {
            runtime.log_info("service:embedding", "Embedding service stopped");
        }
        self.cache.clear();
        self.runtime = None;
        Ok(())
    }
}
